Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipeline.cpp

```diff
@@ -0,0 +1,4371 @@
+
+                shader->SetVectorW(pc.fReg, localCam, 1.f);
+            } break;
+
+            case plPipeConst::kObjPosWorld: {
+                shader->SetVectorW(pc.fReg, GetLocalToWorld().GetTranslate(), 1.f);
+            } break;
+
+            // UNIMPLEMENTED
+            case plPipeConst::kDirLight1:
+            case plPipeConst::kDirLight2:
+            case plPipeConst::kDirLight3:
+            case plPipeConst::kDirLight4:
+            case plPipeConst::kPointLight1:
+            case plPipeConst::kPointLight2:
+            case plPipeConst::kPointLight3:
+            case plPipeConst::kPointLight4:
+            case plPipeConst::kColorFilter:
+            case plPipeConst::kMaxType: {
+                hsAssert(0, "Unimplemented uniform passed to shader");
+            } break;
+        }
+    }
+}
+
+// ISetShaders /////////////////////////////////////////////////////////////////////////////////////
+// Setup to render using the input vertex and pixel shader. Either or both may
+// be nil, in which case the fixed function pipeline is indicated.
+// Any Pipe Constants the non-FFP shader wants will be set here.
+// Lastly, all constants will be set (as a block) for any non-FFP vertex or pixel shader.
+bool plMetalPipeline::ISetShaders(const plMetalVertexBufferRef* vRef, const hsGMatState blendMode, plShader* vShader, plShader* pShader)
+{
+    hsAssert(vShader, "Can't handle programmable passes without vShader");
+    hsAssert(pShader, "Can't handle programmable passes without pShader");
+    plShaderID::ID vertexShaderID = vShader->GetDecl()->GetID();
+    plShaderID::ID fragmentShaderID = pShader->GetDecl()->GetID();
+
+    plMetalDevice::plMetalLinkedPipeline* pipeline = plMetalDynamicMaterialPipelineState(&fDevice, vRef, blendMode.fBlendFlags, vertexShaderID, fragmentShaderID).GetRenderPipelineState();
+    if (fState.fCurrentPipelineState != pipeline->pipelineState) {
+        fDevice.CurrentRenderCommandEncoder()->setRenderPipelineState(pipeline->pipelineState);
+        fState.fCurrentPipelineState = pipeline->pipelineState;
+    }
+
+    if (vShader) {
+        hsAssert(vShader->IsVertexShader(), "Wrong type shader as vertex shader");
+        ISetPipeConsts(vShader);
+
+        plMetalVertexShader* vRef = (plMetalVertexShader*)vShader->GetDeviceRef();
+        if (!vRef) {
+            vRef = new plMetalVertexShader(vShader);
+            hsRefCnt_SafeUnRef(vRef);
+        }
+        if (!vRef->IsLinked())
+            vRef->Link(&fVShaderRefList);
+
+        vRef->ISetConstants(this);
+    }
+
+    if (pShader) {
+        hsAssert(pShader->IsPixelShader(), "Wrong type shader as pixel shader");
+
+        ISetPipeConsts(pShader);
+
+        plMetalFragmentShader* pRef = (plMetalFragmentShader*)pShader->GetDeviceRef();
+        if (!pRef) {
+            pRef = new plMetalFragmentShader(pShader);
+            hsRefCnt_SafeUnRef(pRef);
+        }
```

Comment:
Maybe I'm missing something, but this looks a little suspicious. If we don't have a metal fragment shader, create a new one, then immediately decrement its reference count.... Doesn't that destroy the fragment shader we just created?

Reply:
Looking. This behavior is also in the plDXPipeline at line 9172.

Reply:
Explained on IRC but:

- New is +1.
- vShader quietly takes ownership in the shader and adds a +1
- This if statement releases to give vShader the exclusive ownership.

It's not transparent - but the behavior is correct.